lib/Runtime/Language/PrototypeChainCache.cpp

```diff
@@ -0,0 +1,361 @@
+}
+
+template <typename T>
+bool
+PrototypeChainCache<T>::CheckProtoChain(_In_ RecyclableObject* prototype)
+{
+    Assert(prototype);
+
+    if (T::IsCached(prototype->GetType()))
+    {
+        Assert(CheckProtoChainInternal(prototype));
+        return true;
+    }
+    return CheckProtoChainInternal(prototype);
+}
+
+template <typename T>
+bool
+PrototypeChainCache<T>::CheckProtoChainInternal(RecyclableObject* prototype)
+{
+    Assert(prototype);
+
+    Type *const originalType = prototype->GetType();
+    ScriptContext *const scriptContext = prototype->GetScriptContext();
+
+    bool onlyOneScriptContext = true;
+    TypeId typeId;
+    for (; (typeId = prototype->GetTypeId()) != TypeIds_Null; prototype = prototype->GetPrototype())
+    {
+        if (typeId == TypeIds_Proxy)
+        {
+            return false;
+        }
+        if (prototype->GetScriptContext() != scriptContext)
+        {
+            onlyOneScriptContext = false;
+        }
+        if (!T::CheckObject(prototype))
+        {
+            return false;
+        }
+    }
+
+    if (onlyOneScriptContext)
+    {
+        // See JavascriptLibrary::typesWithNoSpecialPropertiesInItAndPrototypeChain for a description of
+        // this cache. Technically, we could register all prototypes in the chain but this is good enough for now.
+        T::Cache(originalType);
+    }
+
+    return true;
+}
+
+template <typename T>
+void
+PrototypeChainCache<T>::Clear()
+{
+    for (int i = 0; i < this->types->Count(); ++i)
+    {
+        this->cache.ClearType(this->types->Item(i));
+    }
+    this->types->ClearAndZero();
+}
+
+NoSpecialPropertyCache::NoSpecialPropertyCache(_In_ ScriptContext* scriptContext) :
+    scriptContext(scriptContext)
+{
+}
+
+void
+NoSpecialPropertyCache::ClearType(_In_ Type* type)
+{
+    type->SetThisAndPrototypesHaveNoSpecialProperties(false);
+}
+
+bool
+NoSpecialPropertyCache::IsSpecialProperty(PropertyId propertyId)
+{
+    return propertyId == PropertyIds::_symbolToStringTag
+        || propertyId == PropertyIds::_symbolToPrimitive
+        || propertyId == PropertyIds::toString
+        || propertyId == PropertyIds::valueOf;
+}
+
+bool
+NoSpecialPropertyCache::IsSpecialProperty(_In_ PropertyRecord const* propertyRecord)
+{
+    return IsSpecialProperty(propertyRecord->GetPropertyId());
+}
+
+bool
+NoSpecialPropertyCache::IsSpecialProperty(_In_ JavascriptString* propertyString)
+{
+    return BuiltInPropertyRecords::valueOf.Equals(propertyString)
+        || BuiltInPropertyRecords::toString.Equals(propertyString);
+}
+
+// We can handle default initialization of toString/valueOf for Object.prototype,
+// so check if it's that's what is happening
+bool
+NoSpecialPropertyCache::IsDefaultSpecialProperty(
+    _In_ DynamicObject* instance,
+    _In_ JavascriptLibrary* library,
+    PropertyId propertyId)
+{
+    return instance == library->GetObjectPrototype()
+        && ((propertyId == PropertyIds::toString && !library->GetObjectToStringFunction())
```

Comment:
>GetObjectToStringFunction [](start = 62, length = 25)

when can this return nullptr?

Reply:
It will return nullptr in case Object.prototype.toString has not been initialized yet. This is testing if we are materializing Object.prototype.

---
In reply to: [181910993](https://github.com/Microsoft/ChakraCore/pull/4993#discussion_r181910993) [](ancestors = 181910993)